### DISCUSSION TITLE
Allow Super-tab to accept and cycle through suggestions

### DISCUSSION THREAD

**Discussion by marovira** [saghen/blink.cmp#576] [Q&A] [answered]
### Feature Description

In my current config using `nvim.cmp`, I have the following behaviour configured for the tab key:
```lua
                    ["<Tab>"] = cmp.mapping(function(fallback)
                        local col = vim.fn.col(".") - 1

                        if cmp.visible() then
                            cmp.select_next_item()
                        elseif
                            col == 0 or vim.fn.getline("."):sub(col, col):match("%s")
                        then
                            fallback()
                        else
                            cmp.complete()
                        end
                    end, { "i", "s" }),

```

With this setup, my completion works as shown in this video:
https://github.com/user-attachments/assets/1cf37b1d-a115-4167-a160-4abd76acc9ea

As you can see, pressing tab accepts the current suggestion and can also be used to cycle through the remaining suggestions.
From my understanding, the default preset for super-tab in blink doesn't behave like this, instead using the arrow keys to select the next suggestion(s) and using tab to accept. I tried to dig around through the code to see if I could override it to behave like this, but the functions I found don't seem to allow this behaviour. The closest I've gotten is to be able to cycle through the suggestions with Tab, but I can't get the text to auto-insert like in the video above.

Is there a way to do this? Or can this behaviour be added?

**Comment by marovira** [comment]
Just to follow up on this: this has been asked in #6 and it was noted that the behaviour was broken in the release at the time. I've tested this today with the latest version of blink with the following config and it doesn't work as expected:
```lua
        opts = {
            keymap = {
                preset = "enter",
                ["<Tab>"] = { "accept", "select_next", "fallback" },
                ["<S-Tab>"] = { "select_prev", "fallback" },
            },
```

If I swap the order of `accept` and `select_next`, I can get tab to cycle through the suggestions, but:
* The selected option is never inserted,
* There's no way to accept the suggestion.

Has this issue been fixed in the latest releases? Or is it considered broken?

**Comment by naowalrahman** [comment]
When I have `preset = "enter"`, and `["<Tab>"] = { "select_next", "accept", "fallback" }`, I'm able to cycle through the selections and accept with enter. Ideally though, I'd like selecting to automatically insert the suggestion. Any luck? I tried using `preset = "super-tab"` but that didn't seem to do anything. For reference, I'm on LazyVim.

**Reply by marovira** [reply, marked as answer]
I solved it by setting:

```
opts = {
    completion = { list = { selection = "auto_insert"}}} 
} 
    
```

And leaving tab as `"selec_next", "fallback"`.

It's in the full list of options but I had a hard time keeping track of the indentation levels

**Reply by naowalrahman** [reply]
Thanks! For anyone else looking at this, this is the configuration I have:
```lua
return {
    "saghen/blink.cmp",
    opts = function(_, opts)
        opts.completion = { list = { selection = "auto_insert" } }
        opts.keymap = {
            preset = "enter",
            ["<Tab>"] = { "select_next", "fallback" },
            ["<S-Tab>"] = { "select_prev", "fallback" },
        }
    end,
}
```

**Comment by i-am-egg** [comment]
For those finding this in the future, opts.completion.list.selection needs to be a table now, so you may want this config:

```lua
opts = {
  keymap = {
    preset = 'enter',
    ['<Tab>'] = { 'snippet_forward', 'select_next', 'fallback' },
    ['<S-Tab>'] = { 'snippet_backward', 'select_prev', 'fallback' },
  },
  completion = {
    list = {
      selection = {
        preselect = false,
        auto_insert = true,
      },
    },
  },
}
```

Setting `preselect` to `false` means the first press of Tab selects the first suggestion instead of the second - remove it if you want to be able to press enter (or whatever you set it to) to accept the first suggestion. Remove `snippet_[forward|backward]` to allow tabbing through completions while typing a snippet.
